### ISSUE TITLE
[Research] Add Lineax-based JAX linear-solve and implicit-sensitivity validation backend

### ISSUE THREAD

**Issue by googa27** [googa27/finite_element_options#163]
Privacy class: **public/public-synthetic research only**. No client data, proprietary calibration targets, credentials, or private market observations.

## Motivation

`finite_element_options` already has a clear architecture: scikit-fem/SciPy own canonical FEM assembly/solve behavior, while JAX-capable routes must remain optional and may not leak into contracts/core imports. The repository also explicitly names tangent/adjoint solves, optional JAX automatic differentiation, and bump-and-resolve as sensitivity routes. What is missing is a **bounded, independently validated JAX-native linear-solve/sensitivity backend**.

This issue should evaluate [Lineax](https://docs.kidger.site/lineax/) as that backend. Lineax provides JAX linear operators, direct/iterative solves, least-squares solves, factorization/state reuse, PyTree-valued operators/vectors, and AD-compatible inverse operators. It must be adopted only where it reduces custom numerical code **without changing the canonical scikit-fem/SciPy ownership boundary**.

This is a successor research slice to the existing JAX/Bayesian profiles and the closed regime/Diffrax work (#150), not a redesign of the FEM core.

## Mathematical target

For a parameterized discrete FEM system

```math
A(\theta)u(\theta)=b(\theta),
```

the exact first-order tangent satisfies

```math
A(\theta)\,\frac{\partial u}{\partial\theta_k}
=
\frac{\partial b}{\partial\theta_k}
-
\frac{\partial A}{\partial\theta_k}u.
```

For a scalar functional `J(u, θ)`, an adjoint route should use

```math
A(\theta)^T\lambda = \frac{\partial J}{\partial u},
```

and

```math
\frac{dJ}{d\theta_k}
=
\frac{\partial J}{\partial\theta_k}
+
\lambda^T\left(
\frac{\partial b}{\partial\theta_k}
-
\frac{\partial A}{\partial\theta_k}u
\right).
```

The implementation must distinguish three very different claims:

1. differentiating a **JAX-native parameterized discrete operator**;
2. differentiating a JAX representation of an operator **after** scikit-fem/SciPy assembly;
3. differentiating through scikit-fem assembly itself.

Only (1) and a carefully scoped form of (2) belong in this issue. **Do not claim (3).** A NumPy/SciPy-to-JAX conversion is not magically a differentiable assembly graph.

## Library decision to validate

Primary candidate:

- Lineax docs: https://docs.kidger.site/lineax/
- `linear_solve`: https://docs.kidger.site/lineax/api/linear_solve/
- linear operators: https://docs.kidger.site/lineax/api/operators/
- solvers: https://docs.kidger.site/lineax/api/solvers/
- operator tags: https://docs.kidger.site/lineax/api/tags/

Important caveat: Lineax structure tags are declarations, not proofs. Never tag an operator as symmetric/PSD/diagonal/etc. unless repository code has established the property or the fixture mathematically guarantees it. A false tag is a correctness defect, not a performance hint.

Reference theory and implementation background:

- JAX `custom_linear_solve` / implicit linear-solve differentiation semantics: https://docs.jax.dev/en/latest/_autosummary/jax.lax.custom_linear_solve.html
- Giles (2008), *Collected Matrix Derivative Results for Forward and Reverse Mode Algorithmic Differentiation*: https://people.maths.ox.ac.uk/gilesm/files/NA-08-01.pdf
- Giles & Pierce (2000), *An Introduction to the Adjoint Approach to Design*: https://doi.org/10.1007/s001580050176

## Proposed architecture

Do not expose Lineax objects from public core contracts.

Suggested shape after inspecting current module ownership:

```text
finite_element_options/
  sensitivities/
    ... canonical backend-neutral contracts ...
  linear_solvers/
    ... canonical SciPy policies ...
  validation/evidence/
    jax_linear_solve/
      contracts.py          # JSON-safe benchmark/evidence DTOs
      operator_bridge.py    # explicit conversion of validated discrete operators
      lineax_adapter.py     # lazy optional Lineax import
      sensitivity.py        # JAX tangent/adjoint experiments
      benchmarks.py
```

Exact names may change to satisfy the architecture contract and topology/fan-out limits. Prefer extending an existing cohesive package rather than creating cosmetic directories.

Add an isolated optional extra such as `jax-linear` (or extend an existing JAX research profile only if compatibility is demonstrated). The **base wheel must import and test without JAX/Lineax/Equinox**.

### 1. Backend-neutral sensitivity contract

Define immutable inputs/results for:

- operator identity/hash and dimensions;
- dtype/x64 policy;
- parameter vector and lawful parameter domain;
- primal residual norm;
- tangent/adjoint derivative result;
- solver, tolerances, iteration/factorization reuse diagnostics;
- finite-difference/analytic reference error;
- timing and memory evidence;
- convergence/failure state;
- `research_only`, `production_ready=false` until promotion gates pass.

### 2. Validated operator bridge

Support a deliberately small first surface:

- dense/small sparse-equivalent matrices for oracle tests;
- Black-Scholes 1-D semidiscrete mass/stiffness operator where the parameter dependence is explicit;
- an affine operator family when already available, e.g.

```math
A(\theta)=A_0+\sum_k \theta_k A_k.
```

If sparse conversion requires densification, report it and cap problem size. Do not silently make a memory-regressive path look like an accelerator.

### 3. Lineax solver policy

Benchmark appropriate Lineax solvers against repository SciPy references. Solver choice must follow operator evidence rather than hard-coded preference. Include failure paths for singular, ill-conditioned, falsely tagged, and nonconvergent systems.

Test state/factorization reuse for repeated RHS only when the operator is genuinely unchanged.

### 4. Sensitivity validation

For every promoted sensitivity route compare at least:

- JAX/Lineax AD;
- explicit tangent or adjoint formula;
- symmetric finite differences with refinement;
- analytical derivative where one exists.

For Black-Scholes use price/Delta/Vega or parameter sensitivities only where the discrete parameterization and analytical comparator are semantically aligned. Separate **PDE discretization error** from **derivative implementation error**.

### 5. Time stepping

If implicit Euler/Crank–Nicolson is included, explicitly model

```math
(M+\alpha\Delta t K_{n+1})u_{n+1}
=(M-\beta\Delta t K_n)u_n+g_n,
```

and test reverse-mode sensitivity through multiple steps. Record checkpoint/memory scaling. Do not hide an O(N_steps) tape behind a scalar timing result.

## Scientific validation matrix

At minimum include:

| Case | Purpose |
|---|---|
| diagonal/SPD manufactured system | exact primal and derivative oracle |
| nonsymmetric well-conditioned system | prevent SPD-only assumptions |
| near-singular system | failure/conditioning behavior |
| affine Black-Scholes FEM operator | finance-relevant derivative parity |
| repeated RHS / same operator | solver-state reuse evidence |
| time-step refinement | separate discretization and AD errors |

Use deterministic public-synthetic fixtures and x64 where necessary for derivative-convergence studies.

## Acceptance criteria

- [ ] Base install remains free of JAX, Equinox, and Lineax imports/dependencies.
- [ ] Optional profile installs from a clean wheel on the supported Python/JAX version matrix.
- [ ] Primal Lineax and canonical SciPy solutions agree to an accuracy-normalized tolerance on all supported fixtures.
- [ ] Tangent, adjoint, AD, and finite-difference sensitivities converge to the same result where their assumptions match.
- [ ] At least one analytical Black-Scholes-derived oracle is included; numerical derivative error is reported separately from price discretization error.
- [ ] Singular/nonconvergent/ill-conditioned cases return typed failure diagnostics rather than NaNs or misleading prices.
- [ ] No false structural tags are inferred from matrix shape alone.
- [ ] Dense conversion is bounded and visible in diagnostics; no accidental O(n²) memory promotion for sparse production paths.
- [ ] Existing scikit-fem/SciPy regression/convergence tests remain unchanged and green.
- [ ] Architecture docs/contracts, optional-dependency metadata, SBOM/audit evidence, and installed-wheel tests are updated together.
- [ ] A benchmark report states where Lineax is faster/slower, more/less memory intensive, and whether the feature should be promoted, kept experimental, or rejected.

## Non-goals

- Replacing scikit-fem assembly.
- Claiming differentiation through scikit-fem/SciPy assembly.
- Replacing PETSc for distributed sparse HPC.
- GPU claims without actual measured GPU evidence.
- Promoting JAX arrays into public FEM contracts.
- Building a second general linear algebra library inside this repository.

## Follow-on extensions if evidence is positive

1. matrix-free `FunctionLinearOperator` experiments for high-dimensional operators;
2. preconditioned Krylov studies and reusable preconditioner contracts;
3. implicit differentiation of PDE-constrained calibration objectives with Optimistix in a downstream calibration owner;
4. multi-output adjoints for price/Greek/risk functionals;
5. ROM + differentiable online solves using the existing pyMOR evidence boundary;
6. a common sensitivity evidence contract consumed by FD/FEM/Pricing-XVA without sharing implementation internals.

## Advice for the implementation agent

Start from one small manufactured linear system and one existing 1-D Black-Scholes FEM fixture. Do not begin with a generic abstraction. Preserve the SciPy oracle throughout the work. Add architecture tests before optional imports, then numerical parity, then derivatives, then performance. Any benchmark win without matched error is invalid.

## Verification

Use the repository's current `AGENTS.md` gate set, including architecture contract checks, clean-wheel/profile tests, full numerical regression tests, formatting/type checking, privacy scan, dependency audit/SBOM where configured, and `git diff --check`.

A merge may legitimately conclude **“Lineax remains research-only / no promotion”** if accuracy, sparse-memory behavior, compatibility, or maintenance cost fails the gates.
